### PR TITLE
Clean up alternative track pieces and types, remove remnants of researched track pieces

### PR DESCRIPTION
**Pull request by Gymnasiast**

### CODE REVIEW

File: src/openrct2/scenario/scenario.h [hunk 1]
```diff
@@ -132,8 +132,14 @@ typedef struct rct_s6_data {
...
 	uint32 researched_ride_types[8];
 	uint32 researched_ride_entries[8];
-	uint32 researched_track_types_a[128];
-	uint32 researched_track_types_b[128];
+	union {
+		struct {
+			uint32 researched_track_types_a[128];
+			uint32 researched_track_types_b[128];
+		};
+		uint64 researched_track_types[128];
```
Comment (marijnvdwerf) [inline src/openrct2/scenario/scenario.h]: Wouldn't this get you `(researched_track_types_a[i] << 32) | researched_track_types_a[i+1]` as opposed to the expected `(researched_track_types_a[i] << 32) | researched_track_types_b[i]`?

Comment (Gymnasiast) [inline src/openrct2/scenario/scenario.h]: It looks as if it has been done this way in more, like [here](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/ride/ride.h#L114-L120).

**Comment by Gymnasiast** [comment]
```
(Link target) -> 
  libopenrct2.lib(S6Exporter.obj) : error LNK2001: unresolved external symbol "unsigned __int64 const * const RideTypePossibleTrackConfigurations" (?RideTypePossibleTrackConfigurations@@3QB_KB) [C:\projects\openrct2-ject9\src\openrct2-dll\openrct2-dll.vcxproj]
  C:\projects\openrct2-ject9\bin\openrct2.dll : fatal error LNK1120: 1 unresolved externals [C:\projects\openrct2-ject9\src\openrct2-dll\openrct2-dll.vcxproj]
```
Why does it do this? I included track_data.h in S6Exporter.cpp.

File: src/openrct2/scenario/scenario.h (hunk 1 above)
Comment (IntelOrca) [inline src/openrct2/scenario/scenario.h]: You can't do this I'm afraid. You end up interleaving a and b.

**Comment by IntelOrca** [comment]
External C required somewhere. Likely need around one of the headers.

File: src/openrct2/scenario/scenario.h (hunk 1 above)
Comment (marijnvdwerf) [inline src/openrct2/scenario/scenario.h]: The other place only combines two `uint32`s. You're trying to combine two arrays.

Comment (Gymnasiast) [inline src/openrct2/scenario/scenario.h]: Reverted it.

**Comment by Gymnasiast** [comment]
Ok, hopefully this approach will work.

File: src/openrct2/rct2/S6Exporter.cpp [hunk 2]
```diff
@@ -203,7 +206,12 @@ void S6Exporter::Export()
     memcpy(_s6.researched_ride_types, gResearchedRideTypes, sizeof(_s6.researched_ride_types));
     memcpy(_s6.researched_ride_entries, gResearchedRideEntries, sizeof(_s6.researched_ride_entries));
     // Not used by OpenRCT2 any more, but left in to keep RCT2 export working.
-    memcpy(_s6.researched_track_types, RideTypePossibleTrackConfigurations, sizeof(_s6.researched_track_types));
+    for (uint8 i = 0; i < Util::CountOf(RideTypePossibleTrackConfigurations); i++) {
+        researchedTrackPiecesA[i] = (RideTypePossibleTrackConfigurations[i]         ) & 0xFFFFFFFFULL;
+        researchedTrackPiecesB[i] = (RideTypePossibleTrackConfigurations[i] >> 32ULL) & 0xFFFFFFFFULL;
```
Comment (marijnvdwerf) [inline src/openrct2/rct2/S6Exporter.cpp]: Is this ANDing neccesary?

File: src/openrct2/rct2/S6Exporter.cpp [hunk 3]
```diff
@@ -201,8 +205,13 @@ void S6Exporter::Export()
...
     memcpy(_s6.researched_ride_entries, gResearchedRideEntries, sizeof(_s6.researched_ride_entries));
-    memcpy(_s6.researched_track_types_a, gResearchedTrackTypesA, sizeof(_s6.researched_track_types_a));
-    memcpy(_s6.researched_track_types_b, gResearchedTrackTypesB, sizeof(_s6.researched_track_types_b));
+    // Not used by OpenRCT2 any more, but left in to keep RCT2 export working.
+    for (uint8 i = 0; i < Util::CountOf(RideTypePossibleTrackConfigurations); i++) {
+        researchedTrackPiecesA[i] = (RideTypePossibleTrackConfigurations[i]         ) & 0xFFFFFFFFULL;
+        researchedTrackPiecesB[i] = (RideTypePossibleTrackConfigurations[i] >> 32ULL) & 0xFFFFFFFFULL;
+    }
+    memcpy(_s6.researched_track_types_a, researchedTrackPiecesA, sizeof(_s6.researched_track_types_a));
+    memcpy(_s6.researched_track_types_b, researchedTrackPiecesB, sizeof(_s6.researched_track_types_b));
```
Comment (IntelOrca) [inline src/openrct2/rct2/S6Exporter.cpp]: `Memory::Copy`

File: src/openrct2/windows/ride.c [hunk 4]
```diff
@@ -2487,7 +2487,7 @@ static rct_string_id window_ride_get_status_vehicle(rct_window *w, void *argumen
 			trackType == TRACK_ELEM_60_DEG_UP_TO_FLAT ||
 			trackType == TRACK_ELEM_DIAG_25_DEG_UP_TO_FLAT ||
 			trackType == TRACK_ELEM_DIAG_60_DEG_UP_TO_FLAT) {
-			if (track_type_is_invented(ride->type, TRACK_BLOCK_BRAKES) && vehicle->velocity == 0) {
+			if (track_piece_is_available_for_ride_type(ride->type, TRACK_BLOCK_BRAKES) && vehicle->velocity == 0) {
```
Comment (IntelOrca) [inline src/openrct2/windows/ride.c]: I question why it checks for this... is it necessary?

File: src/openrct2/rct2/S6Exporter.cpp (hunk 3 above)
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Exporter.cpp]: The entire file uses memcpy. I can replace all calls if you want me to.

File: src/openrct2/windows/ride.c (hunk 4 above)
Comment (Gymnasiast) [inline src/openrct2/windows/ride.c]: I think it prevent 'Stopped by block brakes' from appearing if the vehicle has just stalled on one of these pieces on a non-block brake ride.

File: src/openrct2/rct2/S6Exporter.cpp (hunk 2 above)
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Exporter.cpp]: I think so. It has been done this way in the old code.

**Comment by Gymnasiast** [comment]
I answered the line notes and rebased the PR.

File: src/openrct2/rct2/S6Exporter.cpp (hunk 3 above)
Comment (IntelOrca) [inline src/openrct2/rct2/S6Exporter.cpp]: Ugh, never mind then

**Pull request merged**